Initial Crossgen2 fixes for composite R2R support

trylek opened the pull request:
With initial local changes to enable composite R2R I have finally
been able to compile a CoreCLR unit test in this mode. It doesn't
yet work at runtime, these changes just fix initial compiler issues
I hit during the build:

1) Delegate constructor signature was missing some comparison logic
that ended up asserting in the node comparer in the composite
build.

2) I have removed two seeminly superfluous parameters to
CreateMethodEntrypointNodeHelper.

3) I have removed an assertion related to multi-file RVA fields
that I believe to be no longer relevant after the refactorings
to copy the RVA fields around properly.

Thanks

Tomas

trylek committed:
Initial Crossgen2 fixes for composite R2R support

With initial local changes to enable composite R2R I have finally
been able to compile a CoreCLR unit test in this mode. It doesn't
yet work at runtime, these changes just fix initial compiler issues
I hit during the build:

1) Delegate constructor signature was missing some comparison logic
that ended up asserting in the node comparer in the composite
build.

2) I have removed two seeminly superfluous parameters to
CreateMethodEntrypointNodeHelper.

3) I have removed an assertion related to multi-file RVA fields
that I believe to be no longer relevant after the refactorings
to copy the RVA fields around properly.

Thanks

Tomas

MichalStrehovsky commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/DelegateCtorSignature.cs:
```diff
@@ -15,18 +15,22 @@ public class DelegateCtorSignature : Signature
         public DelegateCtorSignature(
             TypeDesc delegateType,
             IMethodNode targetMethod,
+            bool isInstantiatingStub,
```
How does this help? It seems like we use `Method.HasInstantiation` at the callsite to populate this, but then we don't really use it for anything (besides when computing the name, which I believe makes the assert you saw go away). `isInstantiatingStub: _targetMethod.Method.HasInstantiation` below is still hardcoded to get the `HasInstantiation` from the method instead of from this new field.

The generic handling here looked pretty suspicious to me so I tried running one of the CoreRT tests on this that I know exercises generics pretty well (it's a test written to exercise AOT compilers, as opposed to the smorgasbord of random tests in the CoreCLR repo that sometimes happen to find AOT bugs). The test fails. I filed #31654.

trylek replied:
You're right, I have verified this is indeed superfluous, it was a leftover from my earlier experiments at fixing the comparison assertion that later turned out to be caused by the incompleteness in CompareToImpl. I have rolled back this part of my change in 2nd commit.

trylek committed:
Address Michal's PR feedback

I have verified locally that the explicit passing of
isInstantiatingStub to DelegateCtorSignature is indeed superflous,
it was a leftover from my previous attempts to fix the aliasing
problem that later turned out to be caused by the incomplete
comparison. I have rolled back this part of my change.

Thanks

Tomas

MichalStrehovsky commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/DelegateCtorSignature.cs:
```diff
@@ -89,7 +89,7 @@ public override int CompareToImpl(ISortableNode other, CompilerComparer comparer
                 return result;
 
-            result = comparer.Compare(_targetMethod.Method, otherNode._targetMethod.Method);
+            result = _targetMethod.CompareToImpl(otherNode._targetMethod, comparer);
```
Wouldn't this work? CompareToImpl is kind of an implementation detail. Calling `Compare` has a couple safeties in it.

```suggestion
            result = comparer.Compare(_targetMethod, otherNode._targetMethod);
```

trylek replied:
Fixed in 3rd commit, thanks for pointing that out.

MichalStrehovsky reviewed (approved):
Looks good otherwise!

trylek committed:
Changed CompareToImpl to Compare per additional Michal's feedback
